### PR TITLE
add support for async hooks

### PR DESCRIPTION
**Pull request by BigMurry**
add support for async hooks as mentioned in #247.

### CODE REVIEW

File: test/hooks-async.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+const request = require('request')
+const sleep = require('then-sleep')
+const fastify = require('..')()
+
+function asyncHookTest (t) {
+  const test = t.test
+  test('hooks - add async preHandler', t => {
+    t.plan(1)
+    try {
+      fastify.addHook('preHandler', async function (request, reply, next) {
```
Comment (mcollina) [inline test/hooks-async.js]: `next`  should not be tampered with if we are returning a promise. They are mutually exclusive, however we don't have any _fast_ means to check this. Let's just assume it's not passed in, ok?

File: test/hooks-async.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        await sleep(200)
```
Comment (mcollina) [inline test/hooks-async.js]: can you make this a `sleep(1)`?

File: docs/Hooks.md [hunk 3]
```diff
@@ -25,6 +25,38 @@ fastify.addHook('onResponse', (res, next) => {
   next()
 })
 ```
+Or `async/await`
+```js
+fastify.addHook('onRequest', async (req, res, next) => {
```
Comment (mcollina) [inline docs/Hooks.md]: can you remove `next` from here and the other instances in this file?

File: test/hooks-async.js (hunk 1 above)
Comment (BigMurry) [reply test/hooks-async.js]: Yes, `next` should not be used inside the async handler.

Comment (mcollina) [reply test/hooks-async.js]: can you remove them from here as well?

Comment (BigMurry) [reply test/hooks-async.js]: It should be.

File: fastify.js [hunk 4]
```diff
@@ -249,7 +249,10 @@ function build (options) {
   }
 
   function hookIterator (fn, cb) {
-    fn(this.req, this.res, cb)
+    const ret = fn(this.req, this.res, cb)
```
Comment (delvedor) [inline fastify.js]: Since this is an hot path, use `var` instead of `const`, it has slightly better performances :)
Same in the others iterators.

File: test/hooks-async.test.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+'use strict'
+
+const tap = require('tap')
```
Comment (delvedor) [inline test/hooks-async.test.js]: You can add this test inside the `hooks.test.js` file :)

File: fastify.js (hunk 4 above)
Comment (BigMurry) [reply fastify.js]: Didn't notice that :)

**Comment by StarpTech** [comment]
Benchmark?

**Comment by delvedor** [comment]
@BigMurry perfect, now we wait for #291 :D

> Benchmark?

Should be almost the same, the change is minimal.

**Comment by StarpTech** [comment]
@BigMurry even `var` is small and has an impact. We should always attach a benchmark to changes in hot paths.

**Comment by BigMurry** [comment]
I cloned [benchmark project](https://github.com/fastify/benchmarks) and changed fastify version to 0.30.2 run benchmark 5 times and then used `yarn link` to change to my branch. Same parameter 5 times, tell me if I'm doing it wrong :) 
**macOS Sierra, MBP 15-inch, Mid 2015, 2.2GHz i7, 16G 1600MHz DDR3**

**Parameters**:

|connection| pipeling|time|
|---|---|---|
|100|10|10|

## my branch
|index|requests average|requests total|
|---|---|---|
|1|31226.91|343475|
|2|31229.82|343637|
|3|29827.64|328121|
|4|32133.6|321311|
|5|32069.6|320688|

## v0.30.2
|index|requests average|requests total|
|---|---|---|
|1|31007.28|341097|
|2|32400.73|356376|
|3|31665.6|316654|
|4|30052.8|300613|
|5|31103.28|342138|

**Comment by StarpTech** [comment]
@BigMurry it's easier to run [this](https://github.com/fastify/fastify/blob/master/package.json#L17) or on windows 

- `npm install -g autocannon`
- Run examples/simple.js
- Run `autocannon -c 100 -d 5 -p 10 localhost:3000/`

**Comment by StarpTech** [comment]
@BigMurry We need always benchmarks of node 4 and node 8.

**Comment by delvedor** [comment]
@StarpTech that's not correct, we run the benchmarks with the current LTS, so Node v6 and v8. Node v4 is in maintenance.

Regarding how take them, it depends, the most important thing it to benchmark your feature branch against the master branch.
You can use the benchmark repo, the script suggested by @StarpTech or [this](https://github.com/fastify/fastify/blob/master/package.json#L17) one written by @lucamaraschi :)

**Comment by StarpTech** [comment]
But then we need still the benchmarks for 6 and 8. @BigMurry What's the current version?

**Comment by delvedor** [comment]
> @delvedor good to know. Is it documented?

Not at the moment, I'll do that asap.

By the way in my opinion is enough benchmark test just Node v8 which is the latest LTS, obviously the result should be similar in Node v6, but except for some edge cases there should be no difference between node v6 and v8 (read if is faster/slower in v8, probably will be faster/slower in node v6).

**Comment by jsumners** [comment]
Erm, no. The performance difference between Node 6 and Node 8 is significant -- https://www.nearform.com/blog/node-js-is-getting-a-new-v8-with-turbofan/

**Comment by delvedor** [comment]
@jsumners I didn't explained myself correctly.
What I meant was: except for some edge cases (the one indicated that article for example) the results are similar. Not in terms of numbers, Node v8 is faster than Node v6, but usually if you go faster in Node v8 you will likely go faster in Node v6.

Take this pr for example, we are not introducing anything that could have different result between Node v6 and v8.

Have I been clearer? :)

**Comment by jsumners** [comment]
Okay.

**Comment by BigMurry** [comment]
> @BigMurry What's the current version?
v8.7.0

**Comment by mcollina** [comment]
@BigMurry can you rebase this on top of master and add support for `onSend`  as well?

**Comment by BigMurry** [comment]
@mcollina Sure, I'll do that.

**Comment by mcollina** [comment]
ahum, the git history does not look right. Can you rebase and make it just a single commit on top of master?

**Comment by BigMurry** [comment]
@mcollina I'll try. :)

**Comment by mcollina** [comment]
Still LGTM.
